query-compiler/query-compiler/src/data_mapper.rs

```diff
@@ -0,0 +1,204 @@
+        let node = graph.node_content(&idx);
+        if let Some(Node::Query(query)) = node {
+            return map_query(query);
+        }
+    }
+
+    for idx in graph.root_nodes() {
+        let node = graph.node_content(&idx);
+        if let Some(Node::Query(query)) = node {
+            return map_query(query);
+        }
+    }
+
+    None
+}
+
+fn map_query(query: &Query) -> Option<ResultNode> {
+    match query {
+        Query::Read(read_query) => map_read_query(read_query),
+        Query::Write(write_query) => map_write_query(write_query),
+    }
+}
+
+fn map_read_query(query: &ReadQuery) -> Option<ResultNode> {
+    match query {
+        ReadQuery::RecordQuery(q) => get_result_node(&q.selected_fields, &q.selection_order, Some(&q.nested)),
+        ReadQuery::ManyRecordsQuery(q) => get_result_node(&q.selected_fields, &q.selection_order, Some(&q.nested)),
+        ReadQuery::RelatedRecordsQuery(q) => get_result_node(&q.selected_fields, &q.selection_order, Some(&q.nested)),
+        ReadQuery::AggregateRecordsQuery(q) => get_result_node_for_aggregation(&q.selectors, &q.selection_order),
+    }
+}
+
+fn map_write_query(query: &WriteQuery) -> Option<ResultNode> {
+    match query {
+        WriteQuery::CreateRecord(q) => get_result_node(&q.selected_fields, &q.selection_order, None),
+        WriteQuery::CreateManyRecords(q) => get_result_node_for_create_many(&q.selected_fields),
+        WriteQuery::UpdateRecord(u) => {
+            match u {
+                UpdateRecord::WithSelection(w) => get_result_node(&w.selected_fields, &w.selection_order, None),
+                UpdateRecord::WithoutSelection(_) => None, // No result data
+            }
+        }
+        WriteQuery::DeleteRecord(q) => get_result_node_for_delete(&q.selected_fields),
+        WriteQuery::UpdateManyRecords(q) => get_result_node_for_update_many(&q.selected_fields),
+        WriteQuery::DeleteManyRecords(_) => None, // No result data
+        WriteQuery::ConnectRecords(_) => None,    // No result data
+        WriteQuery::DisconnectRecords(_) => None, // No result data
+        WriteQuery::ExecuteRaw(_) => None,        // No data mapping
+        WriteQuery::QueryRaw(_) => None,          // No data mapping
+        WriteQuery::Upsert(q) => get_result_node(&q.selected_fields, &q.selection_order, None),
+    }
+}
+
+fn get_result_node(
+    field_selection: &FieldSelection,
+    selection_order: &Vec<String>,
+    nested_queries: Option<&Vec<ReadQuery>>,
+) -> Option<ResultNode> {
+    // println!("field_selection: {field_selection:?}");
+    // println!("selection_order: {selection_order:?}");
+    // println!("nested_queries: {nested_queries:?}");
+
+    // TODO: aggregate-join issue:
+    // selection_order contains "_count"
+    // but the field_map has: "_aggr_count_activations":
+    // Virtual(RelationCount(RelationField("User.activations"), None)), "id": Scalar(ScalarField("User.id"))
+    // There is no nested_queries in this case.
```

Comment:
Debug logging and some information left here for the subsequent work on fixing the second way Prisma is doing aggregations.

Reply:
Removed these for now, less clutter.